### PR TITLE
search: improve literal parentheses interpretation

### PR DESCRIPTION
**Pull request by rvantonder**
Previously a literal pattern like `repo:foo Args or main)` wouldn't work as expected. This was interpreted as `"Args" or (concat "main" ")")`, and concat implies a space separation, meaning `"main )"` instead of `"main)"`. Instead, the intuitive interpretation is that the dangling right parenthesis should be merged above, to be `"main)"`, which is what this PR does.

We cannot always interpret things in this natural way: there simply are ambiguous cases that are currently too tricky to handle once such a pattern is combined with groups, like `(respObj.Size and (data)))` (is the second to last parenthesis trying to close the group, and a typo happened? or is the second pattern intentionally meant to be `(data))`?). For such tricky cases we just return a friendly error message for now. In future we can consider suggestions.

### CODE REVIEW

File: internal/search/query/literal_parser_test.go [hunk 1]
```diff
@@ -41,254 +39,329 @@ func TestParseAndOrLiteral(t *testing.T) {
 		{
 			Input:      "()",
 			Want:       `"()"`,
-			WantLabels: "HeuristicParensAsPatterns",
+			WantLabels: "HeuristicParensAsPatterns,Literal",
```
Comment (rvantonder) [inline internal/search/query/literal_parser_test.go:42]: @unknwon sorry didn't update the struct fields to be private yet--will do in a separate PR soon, no time for cosmetic changes right now :-)

File: internal/search/query/literal_parser.go [hunk 2]
```diff
@@ -144,11 +144,28 @@ loop:
...
+
+				// Heuristic: This right paren may be one we should associate with a previous pattern, and not
+				// just a dangling one. Check if a pattern occured before it and append it if so.
+				if p.pos > 0 {
+					// Heuristic is imprecise and that's OK: It will only look for a 1-byte whitespace
+					// character (not any unicode whitespace) before this paren.
+					if r, _ := utf8.DecodeRune([]byte{p.buf[p.pos-1]}); !unicode.IsSpace(r) {
+						if len(nodes) > 0 {
+							if previous, ok := nodes[len(nodes)-1].(Pattern); ok {
+								previous.Value = previous.Value + pattern.Value
```
Comment (tsenart) [inline internal/search/query/literal_parser.go:160]:
```suggestion
								previous.Value += pattern.Value
```

File: internal/search/query/literal_parser_test.go [hunk 3]
```diff
@@ -24,9 +24,7 @@ func heuristicLabels(nodes []Node) string {
 	labels := Strings(collectLabels(nodes))
 	var keep []string
 	for _, s := range labels {
```
Comment (tsenart) [inline internal/search/query/literal_parser_test.go:26]: Seems like this loop is unnecessary now.

Comment (rvantonder) [reply internal/search/query/literal_parser_test.go]: Good catch!
